### PR TITLE
Added eltype for multivariate distributions

### PR DESCRIPTION
**Pull request by theogf**
Should solve #813

### CODE REVIEW

**Comment by mschauer** [comment]
This conflates `eltype` (e.g. Vector{Float64}) with the `eltype` of the `eltype` (e.g `Float64`).

**Comment by theogf** [comment]
@mschauer So I am not sure what you mean by this, but for `MvNormal`, multiple functions rely on `eltype(d::MvNormal)` and it is currently set to a default `Float64`
I removed the `eltype` for `Multonomial` as they were a problem for the tests.

**Comment by matbesancon** [comment]
@mschauer in that case `eltype` could be the atomic-most type? Another issue is that what is eltype if the type of the covariance matrix and the mean vector differ?

**Comment by theogf** [comment]
> @mschauer in that case `eltype` could be the atomic-most type? Another issue is that what is eltype if the type of the covariance matrix and the mean vector differ?

From the constructors it does not seem possible to have different elementary types for the mean and the covariance (also it would not make so much sense).

**Comment by matbesancon** [comment]
it would if you are doing Automatic Differentiation on the mean, the covariance would be Float while the mean elements would be Dual

**Comment by theogf** [comment]
I am not super familiar with the backend of AD but what would be the negative side of having both Dual?

**Comment by mschauer** [comment]
I see the interface exists a while but I think if there is need for a function returning the atomic element type it should not be called `eltype`. In the case your change is "internally consistent" with the current interface we could  move this concern to a separate issue though.

**Comment by theogf** [comment]
`eltype` is needed since it is called directly by `rand`

**Comment by mschauer** [comment]
Then rand  calls the wrong function ...

**Comment by matbesancon** [comment]
`eltype(::Dirichlet{T}) where {T<:Real} = T`
@theogf you don't need the type bound here, since by construction a Dirichlet has always `T<:Real`

**Comment by matbesancon** [comment]
@mschauer are you good with fixing this behaviour here? As mentioned modifying the whole interface can go in another issue

**Comment by mschauer** [comment]
Ok!

**Comment by matbesancon** [comment]
ping @theogf can you merge master into this branch and adapt for the comment I let above?

**Comment by theogf** [comment]
Sorry I am away till next week, I take care of it as soon as I am back.

**Comment by rfourquet** [comment]
I'm not familiar with this package, but wanted to note that currently, in `Random`, the meaning of `eltype` is extended to mean: for whatever object `x` (not a type), we must have `rand(x) isa eltype(x)`. So @mschauer comments go in the direction of having `Distributions` and `Random` be consistent.

However, we are there is a discussion on whether we should have a function `gentype` instead of `eltype` (`rand(x) isa gentype(x)`), which defaults to `eltype` (e.g. `rand(collection)` returns an element of the collection, so `eltype` makes sense). The idea is that `eltype` for a distribution is not an ideal name, but more importantly, do we want the possibility of `eltype(x) != gentype(x)` ? I'm just putting this here in case you want to chime in at https://github.com/JuliaLang/julia/issues/31968 (it would be great!)

**Comment by mschauer** [comment]
This needs to be rebased. Could someone chime in with a link to a good explanation?

**Comment by matbesancon** [comment]
@mschauer I think this is fine with commit 
d2c36fe ?

**Comment by mschauer** [comment]
I thought something went wrong because the diff of the branch versus master shows 65 files changed.

**Comment by matbesancon** [comment]
ah you're right. @theogf can you from your branch:
1. merge origin/master
2. push

This should do it

**Comment by theogf** [comment]
Sorry I completely messed up things with git, the force-push should have got rid of my mistakes

**Comment by matbesancon** [comment]
no trouble. There seems to be some misses in codecov, can you check if there are things that should be covered?

File: src/multivariate/dirichlet.jl [hunk 1]
```diff
@@ -57,6 +57,7 @@ end
 
 length(d::DirichletCanon) = length(d.alpha)
 
+eltype(::Dirichlet{T}) where T = T
```
Comment (matbesancon) [inline src/multivariate/dirichlet.jl]:
```suggestion
eltype(::Dirichlet{T}) where {T} = T
```

File: src/multivariate/dirichletmultinomial.jl [hunk 2]
```diff
@@ -24,7 +24,6 @@ ntrials(d::DirichletMultinomial) = d.n
 params(d::DirichletMultinomial) = (d.n, d.α)
 @inline partype(d::DirichletMultinomial{T}) where {T<:Real} = T
```
Comment (matbesancon) [inline src/multivariate/dirichletmultinomial.jl]:
```suggestion
@inline partype(d::DirichletMultinomial{T}) where {T} = T
```

File: src/multivariate/mvlognormal.jl [hunk 3]
```diff
@@ -175,6 +175,8 @@ MvLogNormal(Σ::AbstractMatrix) = MvLogNormal(MvNormal(Σ))
 MvLogNormal(σ::AbstractVector) = MvLogNormal(MvNormal(σ))
 MvLogNormal(d::Int,s::Real) = MvLogNormal(MvNormal(d,s))
 
+
+eltype(::MvLogNormal{T}) where {T<:Real} = T
```
Comment (matbesancon) [inline src/multivariate/mvlognormal.jl]:
```suggestion
eltype(::MvLogNormal{T}) where {T} = T
```

File: src/multivariate/mvnormal.jl [hunk 4]
```diff
@@ -219,6 +219,8 @@ MvNormal(Σ::Matrix{<:Real}) = MvNormal(PDMat(Σ))
 MvNormal(σ::Vector{<:Real}) = MvNormal(PDiagMat(abs2.(σ)))
 MvNormal(d::Int, σ::Real) = MvNormal(ScalMat(d, abs2(σ)))
 
+
+eltype(::MvNormal{T}) where {T<:Real} = T
```
Comment (matbesancon) [inline src/multivariate/mvnormal.jl]:
```suggestion
eltype(::MvNormal{T}) where {T} = T
```

**Comment by matbesancon** [comment]
I made suggestions for homogeneity. 
`eltype(::D{T}) where {T}` does not  need `T<:Real` if it is already a type constraint for the distribution

**Comment by mschauer** [comment]
We should at a comment which says what the `eltype` of a distribution is, that is the `eltype` of its samples.

**Comment by theogf** [comment]
@matbesancon I added your suggestions. For the codecov, `eltype` should be called during sampling, I guess there are already tests for that? In any case that should probably be added to the list of tests

@mschauer It is already described this way for the function on the abstract type : https://github.com/JuliaStats/Distributions.jl/blob/432545afd3a29397e2308a99a4978b588676c671/src/common.jl#L51

**Comment by matbesancon** [comment]
> I guess there are already tests for that? In any case that should probably be added to the list of tests

Yes can you check tests and add some? The fact that coverage diff is negative shows that they might not be used.

Also, can you add something in the doc: https://juliastats.github.io/Distributions.jl/stable/univariate/#Common-Interface-1
Adding the doc for `eltype` and `partype` would help people

**Comment by matbesancon** [comment]
ping @theogf

**Comment by theogf** [comment]
So I wrote some tests, but I quickly started to realize that there are a lot more issues. 
 - all `GenericMvTDist` constructors were restricted to Float64 (this is corrected now). 
 - `Float16` cannot be tested since there are no `PDMat{Float16}` constructor for some reason
 - The sampler for `GenerivMvTDist` make calls to the `Gamma` samplers but all `Gamma` structs (`GammIPSampler`, `GammaMTSampler`, etc)are also restricted to `Float64` components. 

Should I just continue pulling the string and parametrizing everything until this works?

File: src/samplers/gamma.jl [hunk 5]
```diff
@@ -6,22 +6,22 @@
...
-    s2::Float64
-    s::Float64
-    i2s::Float64
-    d::Float64
-    q0::Float64
-    b::Float64
-    σ::Float64
-    c::Float64
-    scale::Float64
+struct GammaGDSampler{T} <: Sampleable{Univariate,Continuous}
```
Comment (matbesancon) [inline src/samplers/gamma.jl]: Shouldn't `T` be constrained to `Number` or `Real`?

File: src/samplers/gamma.jl [hunk 6]
```diff
@@ -6,22 +6,22 @@
...
+    b::T
+    σ::T
+    c::T
+    scale::T
 end
 
 function GammaGDSampler(g::Gamma)
     a = shape(g)
-
+    T = typeof(a)
```
Comment (matbesancon) [inline src/samplers/gamma.jl]: `T` could be obtained from the parameter in the signature here

File: test/types.jl [hunk 7]
```diff
@@ -21,3 +22,19 @@ using Distributions
 @assert DiscreteMatrixDistribution <: MatrixDistribution
 @assert ContinuousMatrixDistribution <: ContinuousDistribution
 @assert ContinuousMatrixDistribution <: MatrixDistribution
+
+@testset "Test Sample Type" begin
+    for T in [Float64,Float32,Dual{Nothing,Float64,0}]
+    # for T in [Float64,Float32,Float16,Tracker.TrackedReal{Float64}]
+        @testset "Type $T" begin
+            for d in [MvNormal]#,MvLogNormal,MvNormalCanon,Dirichlet]
```
Comment (matbesancon) [inline test/types.jl]:
```
for d in (MvNormal, )
```
avoids allocating a vector

File: test/types.jl [hunk 8]
```diff
@@ -21,3 +22,19 @@ using Distributions
 @assert DiscreteMatrixDistribution <: MatrixDistribution
 @assert ContinuousMatrixDistribution <: ContinuousDistribution
 @assert ContinuousMatrixDistribution <: MatrixDistribution
+
+@testset "Test Sample Type" begin
+    for T in [Float64,Float32,Dual{Nothing,Float64,0}]
```
Comment (matbesancon) [inline test/types.jl]:
```suggestion
    for T in (Float64,Float32,Dual{Nothing,Float64,0})
```

File: src/samplers/gamma.jl [hunk 9]
```diff
@@ -6,22 +6,22 @@
...
     # Step 1
     s2 = a-0.5
```
Comment (matbesancon) [inline src/samplers/gamma.jl]:
```suggestion
    s2 = a - 0.5
```

File: src/multivariates.jl [hunk 10]
```diff
@@ -14,6 +14,13 @@ Return the sample size of distribution `d`, *i.e* `(length(d),)`.
 """
 size(d::MultivariateDistribution)
 
+
+"""
+    eltype(d::MultivariateDistribution)
```
Comment (matbesancon) [inline src/multivariates.jl]:
```suggestion
    eltype(d::MultivariateDistribution)

```

File: src/multivariate/mvtdist.jl [hunk 11]
```diff
@@ -32,7 +32,10 @@ end
 
 GenericMvTDist(df::Real, μ::Vector{S}, Σ::Cov) where {Cov<:AbstractPDMat, S<:Real} = GenericMvTDist(df, μ, Σ, allzeros(μ))
 
-GenericMvTDist(df::T, Σ::Cov) where {Cov<:AbstractPDMat, T<:Real} = GenericMvTDist(df, zeros(dim(Σ)), Σ, true)
+function GenericMvTDist(df::T, Σ::Cov) where {Cov<:AbstractPDMat, T<:Real}
+    R = Base.promote_eltype(T, Σ)
```
Comment (matbesancon) [inline src/multivariate/mvtdist.jl]: Are you sure `Base.promote_eltype` is what's needed here?

File: src/samplers/gamma.jl (hunk 5 above)
Comment (theogf) [reply src/samplers/gamma.jl]: Indeed! I correct it.

**Comment by matbesancon** [comment]
A significant amount of code has been added, should there be some additions in the doc? In the tests?

File: src/multivariate/mvtdist.jl (hunk 11 above)
Comment (theogf) [reply src/multivariate/mvtdist.jl]: I think so, we want a `promote_type` between `T` and `eltype` of `Σ` and `T<:Real`

**Comment by theogf** [comment]
Most of the code has been about relaxing the constructors `MvTDist` to `Real` variables and to parametrize the `Gamma` samplers (previously fixed to `Float64`).
Maybe I should add a line in `extends.md` to suggest to implement `eltype` for new distributions.

**Comment by matbesancon** [comment]
Yes agreed

**Comment by matbesancon** [comment]
@theogf I'm ok with merging this soon-ish (this week), if you want to add tests before for lines not covered let me know. You can see lines not covered here:
https://codecov.io/gh/JuliaStats/Distributions.jl/compare/432545afd3a29397e2308a99a4978b588676c671...365802f9e4cb438a67adb32b3b1fefe376c7713e/changes

**Comment by matbesancon** [comment]
@theogf sorry I let this slept longer than intended, can you merge master into this branch, I'll merge it then

**Comment by theogf** [comment]
@matbesancon Done! Sorry I also did not answer you, I was first confused by the report and ended up forgetting about it!

**Comment by matbesancon** [comment]
ok, this should be good in ~1hour. I think the coverage decreased because of the new `eltype` methods

**Comment by matbesancon** [comment]
Thanks  @theogf

**Pull request merged**